packages/tools/devtools/devtools-core/src/data-visualization/SharedTreeVisualizer.ts

```diff
@@ -148,64 +155,87 @@ export function toVisualTree(tree: VisualSharedTreeNode): VisualChildNode {
 }
 
 /**
- * Concatenrate allowed types for `ObjectNodeStoredSchema` and `MapNodeStoredSchema`.
+ * Concatenate allowed types for `ObjectNodeStoredSchema` and `MapNodeStoredSchema`.
  */
-function concatenateTypes(fieldTypes: ReadonlySet<string>): string {
+export function concatenateTypes(fieldTypes: ReadonlySet<string>): string {
 	return [...fieldTypes].join(" | ");
 }
 
-/**
- * Returns the allowed fields & types for the object fields (e.g., `foo : string | number, bar: boolean`)
- */
-function getObjectAllowedTypes(schema: SimpleObjectNodeSchema): string {
-	const result: string[] = [];
-
-	for (const [fieldKey, treeFieldSimpleSchema] of Object.entries(schema.fields)) {
-		const fieldTypes = treeFieldSimpleSchema.allowedTypes;
-		result.push(`${fieldKey} : ${concatenateTypes(fieldTypes)}`);
-	}
-
-	return `{ ${result.join(", ")} }`;
-}
-
 /**
  * Returns the schema & fields of the node.
  */
 async function visualizeVerboseNodeFields(
-	tree: VerboseTreeNode,
-	treeSchema: SimpleTreeSchema,
+	treeFields: VerboseTree[] | Record<string, VerboseTree>,
+	treeDefinitions: ReadonlyMap<string, SimpleNodeSchema>,
+	allowedTypes: Record<string, ReadonlySet<string>>,
+	requirements: Record<string, boolean> | undefined,
 	visualizeChildData: VisualizeChildData,
 ): Promise<Record<string, VisualSharedTreeNode>> {
-	const treeFields = tree.fields;
-
 	const fields: Record<string | number, VisualSharedTreeNode> = {};
 
 	for (const [fieldKey, childField] of Object.entries(treeFields)) {
-		fields[fieldKey] = await visualizeSharedTreeNodeBySchema(
+		fields[fieldKey] = await visualizeSharedTreeBySchema(
 			childField,
-			treeSchema,
+			treeDefinitions,
+			allowedTypes[fieldKey],
+			requirements === undefined ? requirements : requirements[fieldKey],
 			visualizeChildData,
 		);
 	}
 
 	return fields;
 }
 
+/**
+ * Extracts and stores allowed types & kind from {@link SimpleFieldSchema}.
+ */
+function storeObjectAllowedTypes(schema: SimpleObjectNodeSchema): {
```

Comment:
Nit: What does "store" mean in this context? I wonder if this should just be something like "getFieldTooltipProperties" or something.

Reply:
Store as just keep the allowed types of the object. I couldn't think of a good name. Changed to `getFieldTooltipProperties`.